xds/internal/balancer/clusterresolver/resource_resolver_dns.go

```diff
@@ -0,0 +1,113 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package clusterresolver
+
+import (
+	"fmt"
+
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+var (
+	newDNS = func(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions) (resolver.Resolver, error) {
+		dnsB := resolver.Get("dns")
+		return dnsB.Build(target, cc, opts)
+	}
+)
+
+// dnsResolver watches updates for the given DNS hostname.
+//
+// It implements resolver.ClientConn interface to work with the DNS resolver.
+type dnsResolver struct {
```

Comment:
What do you think about naming these `dnsDiscoveryMechanism` and `edsDiscoveryMechanism` to indicate that although these use dns and eds to resolve endpoints, they are not a `resolver.Resolver` in the true sense.

In the same vein, the `resolverInterface` in `resource_resolver.go` could be named `discoveryMechanism`?

Reply:
Done.

`discoveryMechanism` can be confusing with the exported `DiscoveryMechanism` (a field in `LBConfig`). But I didn't think of any better names...